controllers/mutation_reconcile_looper.go

```diff
@@ -42,125 +46,209 @@ import (
 	"github.com/open-component-model/ocm-controller/pkg/component"
 	"github.com/open-component-model/ocm-controller/pkg/configdata"
 	"github.com/open-component-model/ocm-controller/pkg/ocm"
+	apiextensionsv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
 )
 
-// tarError defines an error that occurs when the resource is not a tar archive.
-var tarError = errors.New("expected tarred directory content for configuration/localization resources, got plain text")
+// errTar defines an error that occurs when the resource is not a tar archive.
+var errTar = errors.New("expected tarred directory content for configuration/localization resources, got plain text")
 
+// MutationReconcileLooper holds dependencies required to reconcile a mutation object.
 type MutationReconcileLooper struct {
-	Scheme    *runtime.Scheme
-	OCMClient ocm.Contract
-	Client    client.Client
-	Cache     cache.Cache
+	Scheme        *runtime.Scheme
+	OCMClient     ocm.Contract
+	Client        client.Client
+	Cache         cache.Cache
+	DynamicClient dynamic.Interface
 }
 
-func (m *MutationReconcileLooper) ReconcileMutationObject(ctx context.Context, componentVersion *v1alpha1.ComponentVersion, spec v1alpha1.MutationSpec, obj client.Object) (string, error) {
-	log := log.FromContext(ctx)
-	var (
-		resourceData []byte
-		err          error
-	)
+// ReconcileMutationObject reconciles mutation objects and writes a snapshot to the cache.
+func (m *MutationReconcileLooper) ReconcileMutationObject(ctx context.Context, obj v1alpha1.MutationObject) error {
+	mutationSpec := obj.GetSpec()
 
-	if spec.Source.SourceRef == nil && spec.Source.ResourceRef == nil {
-		return "",
-			fmt.Errorf("either sourceRef or resourceRef should be defined, but both are empty")
+	sourceData, err := m.getData(ctx, &mutationSpec.SourceRef)
```

Comment:
This also suffers from re-queueing on failure. Can you create a typed error for snapshot not ready too in this kind and requeue in case of that?

Reply:
ah, even better